xdsclient: update watcher API as per gRFC A88

purnesh42H opened the pull request:
This is the first part of implementing gRFC A88 (https://github.com/grpc/proposal/pull/466).

This introduces the new watcher API. This table summarizes the API changes and behavior for each case:

Case | Old API | New API | Behavior
------- | ---------- | ------------ | ------------
Resource timer fires | `OnResourceDoesNotExist()` | `OnResourceChanged(NOT_FOUND)` | Fail data plane RPCs
LDS or CDS resource deletion | `OnResourceDoesNotExist()` | `OnResourceChanged(NOT_FOUND)` | Drop resource and fail data plane RPCs
xDS channel reports TRANSIENT_FAILURE | `OnError()` | `OnResourceChanged(status)` if resource NOT already cached; `OnAmbientError(status)` otherwise | Continue using cached resource, if any; otherwise, fail data plane RPCs
ADS stream terminates without receiving a response | `OnError()` | `OnResourceChanged(status)` if resource NOT already cached; `OnAmbientError(status)` otherwise | Continue using cached resource, if any; otherwise, fail data plane RPCs
Invalid resource update (client NACK) | `OnError()` | `OnResourceChanged(status)` if resource NOT already cached; `OnAmbientError(status)` otherwise | Continue using cached resource, if any; otherwise, fail data plane RPCs
Valid resource update | `OnUpdate(resource)` | `OnResourceChanged(resource)` | use the new resource

RELEASE NOTES: None

[google3 import run](https://critique.corp.google.com/cl/746298247/analysis) analysis did not show any issues

easwars commented on xds/internal/xdsclient/xdsresource/resource_type.go:
```diff
@@ -58,27 +58,48 @@ type Producer interface {
 
-// ResourceWatcher wraps the callbacks to be invoked for different events
-// corresponding to the resource being watched.
+// ResourceWatcher is an interface that can to be implemented to wrap the
+// callbacks to be invoked for different events corresponding to the resource
+// being watched.
```
I don't feel like this change to the comment adds much value. The fact that it is an interface is clear from the type definition and since it is an interface, it *needs *to be implemented. So, I would leave this unchanged or improve it in a different way.

easwars commented on xds/internal/xdsclient/xdsresource/resource_type.go:
```diff
@@ -58,27 +58,48 @@ type Producer interface {
-	// OnUpdate is invoked to report an update for the resource being watched.
+	// OnResourceChanged is invoked to notify the watcher of a new version of
+	// the resource received from the xDS server or an error indicating the
+	// reason why the resource cannot be obtained.
```
Nit: s/cannot/could not/

easwars commented on xds/internal/xdsclient/xdsresource/resource_type.go:
```diff
@@ -58,27 +58,48 @@ type Producer interface {
-	OnUpdate(ResourceData, OnDoneFunc)
-
-	// OnError is invoked under different error conditions including but not
+	// type for the resource being watched. In case of error, the ResourceData
+	// is nil otherwise its not nil and error is nil but both will never be nil
+	// together.
```
This is super confusing.

Should we consider implementing a naive version of the C++ `StatusOr`?

It could be something like
```golang

type ResourceDataOrError {
  data ResourceData
  err error
}

func (r *ResourceDataOrError) ResourceData {
  return r.data
}

func (r *ResourceDataOrError) Error {
  return r.err
}

func NewResourceDataOrErrorFromResourceData(data ResourceData) *ResourceDataOrError {
  return &ResourceDataOrError{data: data}
}

func NewResourceDataOrErrorFromError(err error) *ResourceDataOrError {
  return &ResourceDataOrError{err: err}
}
```

The fact that the fields are unexported means that the above struct can only be constructed with one of the above `NewXxx`, and that guarantees that only one of the two fields will be set at any given point in time.

The names of the `NewXxx` functions suck at this point. But maybe with some thought, we can come up with something better.

Or we could event come up with a generic version of this which takes inspiration from the Rust `Result<T, E>` type.

@dfawley : thoughts?

dfawley replied:
I think the Go-typical way to implement a sum type is to just have two fields, only one of which is set.  And a comment saying as much.

```go
// Whatever contains either Data or Err if an error occurred while trying to obtain Data.
type Whatever struct {
	Err error
	Data Something
}
```

That effectively mirrors:

```go
func SomethingThatCouldFail() (Something, error) {}
```

purnesh42H replied:
Added

```
// ResourceDataOrError is a struct that contains either ResourceData or error.
// It is used to represent the result of an xDS resource update. Exactly one of
// Data or Err will be non-nil.
type ResourceDataOrError struct {
	Data ResourceData
	Err  error
}
```

easwars commented on xds/internal/xdsclient/xdsresource/resource_type.go:
```diff
@@ -58,27 +58,55 @@ type Producer interface {
+	// will never be nil together.
+	//
+	// Watcher is expected to use the most recent value passed to
+	// OnResourceChanged(), regardless of whether that's a resource or an error
```
Nit: s/ whether that's a resource/whether that's a ResourceData/

purnesh42H replied:
Done

easwars commented on xds/internal/xdsclient/xdsresource/resource_type.go:
```diff
@@ -58,27 +58,55 @@ type Producer interface {
-	OnUpdate(ResourceData, OnDoneFunc)
-
-	// OnError is invoked under different error conditions including but not
+	// OnResourceChanged is invoked to notify the watcher of a new version of
+	// the resource received from the xDS server or an error indicating the
+	// reason why the resource could not be obtained.
+	//
+	// The ResourceData of the ResourceDataOrError needs to be type asserted to
+	// the appropriate type for the resource being watched. In case of error,
+	// the ResourceData is nil otherwise its not nil and error is nil but both
+	// will never be nil together.
```
How about replacing the first two paragraphs with the following:

```
	// OnResourceChanged is invoked to notify the watcher of a new version of
	// the resource received from the xDS server or an error indicating the
	// reason why the resource could not be obtained. 
	//
	// In the former case, this callback will be invoked with a non-nil ResourceData
	// in ResourceDataOrError. The ResourceData of the ResourceDataOrError needs
	// to be type asserted to the appropriate type for the resource being watched. 
	//
	// In the latter case, this callback will be invoked with a non-nil error value in ResourceDataOrError.
```

purnesh42H replied:
Done

easwars commented on xds/internal/xdsclient/xdsresource/cluster_resource_type.go:
```diff
@@ -110,39 +110,42 @@ func (c *ClusterResourceData) Raw() *anypb.Any {
+	//      - connection failure (if resource is not cached)
+	OnResourceChanged(*ResourceDataOrError, OnDoneFunc)
+
+	// If resource is already cached, it is invoked under different error
```
Please start with the docstring with the name of the function, as is customary in Go.

purnesh42H replied:
Done

easwars commented on xds/internal/xdsclient/xdsresource/endpoints_resource_type.go:
```diff
@@ -106,39 +106,42 @@ func (e *EndpointsResourceData) Raw() *anypb.Any {
+	//      - connection failure (if resource is not cached)
+	OnResourceChanged(*ResourceDataOrError, OnDoneFunc)
+
+	// If resource is already cached, it is invoked under different error
```
Please start with the docstring with the name of the function, as is customary in Go.

purnesh42H replied:
Done

easwars commented on xds/internal/xdsclient/xdsresource/listener_resource_type.go:
```diff
@@ -143,39 +143,42 @@ func (l *ListenerResourceData) Raw() *anypb.Any {
+	//      - connection failure (if resource is not cached)
+	OnResourceChanged(*ResourceDataOrError, OnDoneFunc)
+
+	// If resource is already cached, it is invoked under different error
```
Please start with the docstring with the name of the function, as is customary in Go.

easwars commented on xds/internal/xdsclient/xdsresource/route_config_resource_type.go:
```diff
@@ -107,39 +107,42 @@ func (r *RouteConfigResourceData) Raw() *anypb.Any {
+	//      - connection failure (if resource is not cached)
+	OnResourceChanged(*ResourceDataOrError, OnDoneFunc)
+
+	// If resource is already cached, it is invoked under different error
```
Please start with the docstring with the name of the function, as is customary in Go.

purnesh42H replied:
Done

easwars commented on xds/internal/xdsclient/authority.go:
```diff
@@ -455,17 +457,17 @@ func (a *authority) handleADSResourceUpdate(serverConfig *bootstrap.ServerConfig
 			watcherCnt.Add(1)
-			funcsToSchedule = append(funcsToSchedule, func(context.Context) { watcher.OnResourceDoesNotExist(done) })
+			funcsToSchedule = append(funcsToSchedule, func(context.Context) {
+				watcher.OnResourceChanged(xdsresource.ResourceDataOrError{Err: xdsresource.NewErrorf(xdsresource.ErrorTypeResourceNotFound, "xds: resource has been removed")}, done)
```
I think we should include the resource name in the error for it to be more useful.

purnesh42H replied:
Added

easwars commented on xds/internal/xdsclient/authority.go:
```diff
@@ -507,7 +509,9 @@ func (a *authority) handleADSResourceDoesNotExist(rType xdsresource.Type, resour
 		watcher := watcher
-		a.watcherCallbackSerializer.TrySchedule(func(context.Context) { watcher.OnResourceDoesNotExist(func() {}) })
+		a.watcherCallbackSerializer.TrySchedule(func(context.Context) {
+			watcher.OnResourceChanged(xdsresource.ResourceDataOrError{Err: xdsresource.NewErrorf(xdsresource.ErrorTypeResourceNotFound, "xds: resource %s does not exist", rType.TypeName())}, func() {})
```
Same here. I think we should include the resource name in the error for it to be more useful.

purnesh42H replied:
Added

easwars commented on xds/internal/xdsclient/tests/resource_update_test.go:
```diff
@@ -161,7 +161,7 @@ func (s) TestHandleListenerResponseFromManagementServer(t *testing.T) {
 				}},
 			},
-			wantErr: "Listener not found in received response",
+			wantErr: "xds: resource ListenerResource does not exist",
```
I'm thinking if it makes sense to split out these tests that verify the "resource-not-found" behavior. The advantage of splitting them out is that the test logic can instead verify that the error type is `xdsresource.ErrorTypeResourceNotFound` instead of verifying the error string.

This change will apply to tests for all resource types in this file.

purnesh42H replied:
So, each resource type has 3 test cases of resource does not exist which will change. So, not sure how much benefit it will be to move them in the single test. There will be lot of if else or switch cases etc. as well. Also, may be its fine to assert on error message as that is the error message that we are going to send downstream. For now i just modified the tests cases wants to include resource name. Let me know if you feel strongly to change this.

easwars commented on xds/internal/testutils/resource_watcher.go:
```diff
@@ -35,19 +35,36 @@ type TestResourceWatcher struct {
+		}
+		w.ErrorCh <- update.Err
+		return
+
```
Nit: remove unnecessary newline.

purnesh42H replied:
Done

easwars commented on xds/internal/xdsclient/xdsresource/listener_resource_type.go:
```diff
@@ -143,39 +143,42 @@ func (l *ListenerResourceData) Raw() *anypb.Any {
+	//      - resource validation error (if resource is not cached)
+	//      - ADS stream failure (if resource is not cached)
+	//      - connection failure (if resource is not cached)
+	OnResourceChanged(*ResourceDataOrError, OnDoneFunc)
```
Why is the `ResourceDataOrError` being passed as a pointer in the resource type specific watchers, but is passed as value in the resource-agnostic `ResourceWatcher` interface? I'm guessing you just followed the existing pattern. I think we should revisit this after this PR. There is a TODO in these files about this. But I'm OK for it to be the way it currently is for this PR.

purnesh42H replied:
yeah i followed the existing pattern. There is TODO about considering using pointer for ListenerUpdate. IS that what you are referring to? or should we add another TODO here?

easwars commented on xds/internal/server/rds_handler.go:
```diff
@@ -156,26 +156,22 @@ func (rw *rdsWatcher) OnUpdate(update *xdsresource.RouteConfigResourceData, onDo
 	if rw.logger.V(2) {
-		rw.logger.Infof("RDS watch for resource %q received update: %#v", rw.routeName, update.Resource)
+		if update.Err != nil {
+			rw.logger.Infof("RDS watch for resource %q received error: %#v", rw.routeName, update.Err)
```
Why do we need the `%#v` format directive here? `%v` format directive has usually been sufficient for outputting errors.

purnesh42H replied:
yeah. Not needed. They were already at 3 places. Updated them.

easwars commented on xds/internal/server/rds_handler.go:
```diff
@@ -156,26 +156,22 @@ func (rw *rdsWatcher) OnUpdate(update *xdsresource.RouteConfigResourceData, onDo
 	rw.mu.Unlock()
 	if rw.logger.V(2) {
-		rw.logger.Infof("RDS watch for resource %q received update: %#v", rw.routeName, update.Resource)
+		if update.Err != nil {
```
I think this code block should be restructured as follows:

```golang
func (rw *rdsWatcher) OnResourceChanged(update *xdsresource.ResourceDataOrError, onDone xdsresource.OnDoneFunc) {
	...
	if update.Err != nil {	
		if rw.logger.V(2) {
			// Log the error
		}
		// Handle the error
		return
	}
	
	// Log the update with verbosity check
	// Handle the update.
}
```

purnesh42H replied:
Updated. Thanks.

easwars commented on xds/internal/server/listener_wrapper.go:
```diff
@@ -414,19 +414,33 @@ type ldsWatcher struct {
 	}
 	if lw.logger.V(2) {
-		lw.logger.Infof("LDS watch for resource %q received update: %#v", lw.name, update.Resource)
+		if update.Err != nil {
+			lw.logger.Infof("LDS watch for resource %q received error: %#v", lw.name, update.Err)
+		} else {
+			u := update.Data.(*xdsresource.ListenerResourceData)
+			lw.logger.Infof("LDS watch for resource %q received update: %#v", lw.name, u.Resource)
+		}
+	}
```
Same comment as in the other file. Please add logs along with the code that handles the error or the update instead of having a separate block for logging.

easwars commented on xds/internal/resolver/xds_resolver.go:
```diff
@@ -518,17 +518,21 @@ func (r *xdsResolver) onListenerResourceUpdate(update xdsresource.ListenerUpdate
+func (r *xdsResolver) onListenerResourceChangedError(err error) {
 	if r.logger.V(2) {
-		r.logger.Infof("Received resource-not-found-error for Listener resource %q", r.ldsResourceName)
+		if xdsresource.ErrType(err) == xdsresource.ErrorTypeResourceNotFound {
+			r.logger.Infof("Received resource-not-found-error for Listener resource %q", r.ldsResourceName)
+		} else {
+			r.logger.Infof("Received on-resource-changed error for Listener resource %q: %v", r.ldsResourceName, err)
+		}
```
How about simply printing the error in the log instead of checking for the error type and printing different log statements. Printing the error value itself should give information about what type of error was recieved.

purnesh42H replied:
Just kept the statement in else part

easwars commented on xds/internal/resolver/xds_resolver.go:
```diff
@@ -559,17 +563,21 @@ func (r *xdsResolver) onRouteConfigResourceUpdate(name string, update xdsresourc
+func (r *xdsResolver) onRouteConfigResourceChangedError(name string, err error) {
 	if r.logger.V(2) {
-		r.logger.Infof("Received resource-not-found-error for RouteConfiguration resource %q", name)
+		if xdsresource.ErrType(err) == xdsresource.ErrorTypeResourceNotFound {
+			r.logger.Infof("Received resource-not-found-error for RouteConfiguration resource %q", name)
+		} else {
+			r.logger.Infof("Received on-resource-changed error for RouteConfiguration resource %q: %v", name, err)
+		}
```
Same here.

purnesh42H replied:
Just kept the statement in else part

easwars commented on xds/internal/balancer/clusterresolver/resource_resolver_eds.go:
```diff
@@ -76,20 +76,43 @@ func newEDSResolver(nameToWatch string, producer xdsresource.Producer, topLevelR
 
+	if update.Err != nil {
+		if er.logger.V(2) {
+			if xdsresource.ErrType(update.Err) == xdsresource.ErrorTypeResourceNotFound {
+				er.logger.Infof("EDS discovery mechanism for resource %q reported resource-does-not-exist error", er.nameToWatch)
+			} else {
+				er.logger.Infof("EDS discovery mechanism for resource %q reported on resource changed error: %v", er.nameToWatch, update.Err)
+			}
```
Same here. Please print the error to logs instead of having to check for the type of error.

purnesh42H replied:
Just kept the statement in else part

easwars commented on xds/internal/balancer/cdsbalancer/cdsbalancer.go:
```diff
@@ -525,15 +525,14 @@ func (b *cdsBalancer) onClusterError(name string, err error) {
 // Only executed in the context of a serializer callback.
-func (b *cdsBalancer) onClusterResourceNotFound(name string) {
-	err := xdsresource.NewErrorf(xdsresource.ErrorTypeResourceNotFound, "resource name %q of type Cluster not found in received response", name)
+func (b *cdsBalancer) onClusterResourceChangedError(name string, err error) {
```
Log the error here?

purnesh42H replied:
Added error log

purnesh42H replied:
Actually i switched to warning because error log was failing tests

easwars commented on xds/internal/balancer/cdsbalancer/cdsbalancer.go:
```diff
@@ -525,15 +525,14 @@ func (b *cdsBalancer) onClusterError(name string, err error) {
 		b.ccw.UpdateState(balancer.State{
 			ConnectivityState: connectivity.TransientFailure,
-			Picker:            base.NewErrPicker(err),
+			Picker:            base.NewErrPicker(fmt.Errorf("%q: %v", name, err)),
```
Have you verified that this error is as descriptive as what we had earlier?

purnesh42H replied:
We are just prepending the name to the error. The actual error message is still there. What exactly your doubt is?

purnesh42H commented on xds/internal/xdsclient/xdsresource/resource_type.go:
```diff
@@ -58,27 +58,58 @@ type Producer interface {
+	//      - resource name parsing error
+	//      - resource validation error (if resource is not cached)
+	//      - ADS stream failure (if resource is not cached)
+	//      - connection failure (if resource is not cached)
```
I have removed it. I have mentioned gRFC A88 at the top to refer for exhaustive list. I saw other places in code where we just mention the grfc and don't link. Also, the proposal is not merged yet.

easwars replied:
Sounds good.

purnesh42H commented on xds/internal/xdsclient/xdsresource/resource_type.go:
```diff
@@ -58,27 +58,58 @@ type Producer interface {
+	// conditions including but not limited to the following:
+	//      - resource validation error
+	//      - ADS stream failure
+	//      - connection failure
```
I have removed it. I have mentioned gRFC A88 at the top to refer for exhaustive list. I saw other places in code where we just mention the grfc and don't link. Also, the proposal is not merged yet.

easwars replied:
The language from the gFRC is much stricter than yours:
```
In particular, the watcher should not stop using the previously seen resource, and the XdsClient will not remove the resource from its cache
```

So, please mention that the `watcher should not stop using the previously seen resource` and not `the watcher may continue using the previously seen resource`

purnesh42H replied:
Thanks. Updated.

purnesh42H commented on xds/internal/xdsclient/xdsresource/resource_type.go:
```diff
@@ -58,27 +58,58 @@ type Producer interface {
 // from the xDS server.
 type OnDoneFunc func()
 
+// ResourceDataOrError is a struct that contains either ResourceData or error.
```
Done

purnesh42H commented:
@easwars ptal

easwars commented on xds/internal/balancer/cdsbalancer/cdsbalancer.go:
```diff
@@ -525,15 +525,16 @@ func (b *cdsBalancer) onClusterError(name string, err error) {
 // Only executed in the context of a serializer callback.
-func (b *cdsBalancer) onClusterResourceNotFound(name string) {
-	err := xdsresource.NewErrorf(xdsresource.ErrorTypeResourceNotFound, "resource name %q of type Cluster not found in received response", name)
+func (b *cdsBalancer) onClusterResourceChangedError(name string, err error) {
```
The comment at the top of this method needs to be updated. This method now handles not only resource-not-found errors, but other types of errors as well that instruct the LB policy to stop using the previously received resource.

purnesh42H replied:
Updated. Also, updated the onClusterAmbient error to not stop using the previously seen resource

easwars commented on xds/internal/xdsclient/authority.go:
```diff
@@ -194,7 +194,7 @@ func (a *authority) handleADSStreamFailure(serverConfig *bootstrap.ServerConfig,
 				watcher := watcher
 				a.watcherCallbackSerializer.TrySchedule(func(context.Context) {
-					watcher.OnError(xdsresource.NewErrorf(xdsresource.ErrorTypeConnection, "xds: error received from xDS stream: %v", err), func() {})
+					watcher.OnAmbientError(xdsresource.NewErrorf(xdsresource.ErrorTypeConnection, "xds: error received from xDS stream: %v", err), func() {})
```
Based on the PR description, for the following two cases:
- xDS channel reports TRANSIENT_FAILURE | 
- ADS stream terminates without receiving a response
the old behavior was to call `OnError()`, while the new behavior is to call `OnResourceChanged(status)` if resource NOT already cached, and `OnAmbientError(status)` otherwise.

But, we are only calling the latter here. Am I missing something?

purnesh42H replied:
yeah so, since we have already checked for `ErrTypeStreamFailedAfterRecv` above, the only case left is un cached which means this should always be `OnResourceChanged()`?

@easwars

easwars replied:
`ErrTypeStreamFailedAfterRecv` just means that *some* message was received on the stream before it broke. It does not say anything specific to the resource under question here.

purnesh42H replied:
oh i see. I thought `ErrTypeStreamFailedAfterRecv` meant to state that stream failure happened but we have received the resource once before.

easwars commented:
This came up during some other discussion with @dfawley.
1. Callback methods named `OnXxx` are more of a C++ style. In Go, we would probably just name them `Xxx`.
2. Whether we should have two callback methods, one named `ResourceChanged(ResourceData)`, and the other named `ResourceError(error)`. This is to work around the fact that we don't have a `StatusOr` type in Go, and without that, having two APIs is better than one API (where we need to add a new struct to hold one of two possible values, with no guarantees that only one of the two values is set).

purnesh42H commented:
> This came up during some other discussion with @dfawley.
> 
> 1. Callback methods named `OnXxx` are more of a C++ style. In Go, we would probably just name them `Xxx`.
> 2. Whether we should have two callback methods, one named `ResourceChanged(ResourceData)`, and the other named `ResourceError(error)`. This is to work around the fact that we don't have a `StatusOr` type in Go, and without that, having two APIs is better than one API (where we need to add a new struct to hold one of two possible values, with no guarantees that only one of the two values is set).

I will wait for https://github.com/grpc/grpc-go/pull/8042/ before making further changes to this PR

purnesh42H commented:
@easwars I have made the changes to match the final a88 watcher interface `ResourceChanged`, `ResourceError` and `AmbientError` including the conditions to call `ResourceError` or `AmbientError` based on cache.

Everything else is working fine but looks like this `TestServeAndCloseDoNotRace` is causing race https://github.com/grpc/grpc-go/actions/runs/14091957805/job/39470559995?pr=7977. Server is stopping in the middle of the test. I can't seem to figure how the watcher change is affecting this. Could you help take a look? Thanks.

purnesh42H commented:
> @easwars I have made the changes to match the final a88 watcher interface `ResourceChanged`, `ResourceError` and `AmbientError` including the conditions to call `ResourceError` or `AmbientError` based on cache.
> 
> Everything else is working fine but looks like this `TestServeAndCloseDoNotRace` is causing race https://github.com/grpc/grpc-go/actions/runs/14091957805/job/39470559995?pr=7977. Server is stopping in the middle of the test. I can't seem to figure how the watcher change is affecting this. Could you help take a look? Thanks.

Ah, look like it was failing because of calling ResourceError when channel not found which was calling ListenerWrapper onLDSResourceError which was switching serving mode. Seems like that's unexpected behavior? Calling AmbientError fixed it. Is it correct to consider channel not found to be AmbientError?

purnesh42H commented:
> > @easwars I have made the changes to match the final a88 watcher interface `ResourceChanged`, `ResourceError` and `AmbientError` including the conditions to call `ResourceError` or `AmbientError` based on cache.
> > Everything else is working fine but looks like this `TestServeAndCloseDoNotRace` is causing race https://github.com/grpc/grpc-go/actions/runs/14091957805/job/39470559995?pr=7977. Server is stopping in the middle of the test. I can't seem to figure how the watcher change is affecting this. Could you help take a look? Thanks.
> 
> Ah, look like it was failing because of calling ResourceError when channel not found in authority.go which was calling ListenerWrapper onLDSResourceError which was switching serving mode. Seems like that's unexpected behavior? Calling AmbientError fixed it. Is it correct to consider channel not found to be AmbientError while registering the watcher?

Actually, I thought over it and right thing to do here is handle this case in listener wrapper. Like previously done, we handle lds error only in case of resource-not-found. Let me know if that's okay.

markdroth commented:
> Ah, look like it was failing because of calling ResourceError when channel not found in authority.go which was calling ListenerWrapper onLDSResourceError which was switching serving mode. Seems like that's unexpected behavior? Calling AmbientError fixed it. Is it correct to consider channel not found to be AmbientError while registering the watcher?

If we start a watch for a resource and the authority of that resource is not found in the bootstrap config, then the right behavior is to report a non-ambient error to the watcher.

In general, we should never call OnAmbientError() for a watcher before we call OnResourceChanged().  If we get a transient error before we've seen a valid resource, then the transient error should be reported to OnResourceChanged() instead of OnAmbientError().

purnesh42H commented:
> > Ah, look like it was failing because of calling ResourceError when channel not found in authority.go which was calling ListenerWrapper onLDSResourceError which was switching serving mode. Seems like that's unexpected behavior? Calling AmbientError fixed it. Is it correct to consider channel not found to be AmbientError while registering the watcher?
> 
> If we start a watch for a resource and the authority of that resource is not found in the bootstrap config, then the right behavior is to report a non-ambient error to the watcher.
> 
> In general, we should never call OnAmbientError() for a watcher before we call OnResourceChanged(). If we get a transient error before we've seen a valid resource, then the transient error should be reported to OnResourceChanged() instead of OnAmbientError().

yeah handled it in ListenerWrapper. Still calling ResourceError for channel not found

markdroth commented:
> yeah handled it in ListenerWrapper

I haven't looked at the code, but that seems wrong to me.  There should not need to be any special handling for individual resource types -- the logic for this in XdsClient should be completely resource-type-agnostic.

The only thing that should be handled in individual resource types are things that are different from other resource types, like determining what resource type to pass to the watcher.

purnesh42H commented:
> > yeah handled it in ListenerWrapper
> 
> I haven't looked at the code, but that seems wrong to me. There should not need to be any special handling for individual resource types -- the logic for this in XdsClient should be completely resource-type-agnostic.
> 
> The only thing that should be handled in individual resource types are things that are different from other resource types, like determining what resource type to pass to the watcher.

This error for channel not found while registering watch https://github.com/grpc/grpc-go/pull/7977/files#diff-b4a775482793516a8ed23bd6a58a805d316f5abb6468580691c2b0556e22606aR619 was recently added. Listener wrapper for grpc xds server had a logic to stop serving only if error is of type ResourceNotFound so i just kept that instead of stopping always https://github.com/grpc/grpc-go/pull/7977/files#diff-e4706c72ae912399b7f8ee6f04cec2374ef7a7679b12358f201ddb0b45e34146R453-R457. If we don't do that `TestServeAndCloseDoNotRace` is causing the race because we are creating and stopping grpc xds servers in a loop . @easwars do you think we should rewrite this test?

markdroth commented:
> Listener wrapper for grpc xds server had a logic to stop serving only if error is of type ResourceNotFound so i just kept that instead of stopping always https://github.com/grpc/grpc-go/pull/7977/files#diff-e4706c72ae912399b7f8ee6f04cec2374ef7a7679b12358f201ddb0b45e34146R453-R457.

I don't think that check is correct -- it should be removed.  If the XdsClient delivers an error, it should always be passed through to the watcher.

That's an inherent part of the API change we're making in this PR.  The errors that we were previously ignoring in `OnError()` are now being passed to `OnAmbientError()` instead.  If you are seeing an error reported to `OnError()`, that means we actually want to drop the resource we saw previously, if any.

easwars commented on xds/internal/balancer/cdsbalancer/cdsbalancer_test.go:
```diff
@@ -813,7 +813,7 @@ func (s) TestResolverError(t *testing.T) {
 	// Grab the wrapped connection from the listener wrapper. This will be used
 	// to verify the connection is closed.
-	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout*100000)
```
Please undo this change.

Just FYI: In a large PR, things like this can be easily missed during the review process. Please be a little more mindful of such changes going forward.

purnesh42H replied:
yeah removed. Will be careful.

easwars commented on xds/internal/xdsclient/metrics_test.go:
```diff
@@ -76,9 +76,9 @@ func (s) TestResourceUpdateMetrics(t *testing.T) {
-			"server_uri": %q,
-			"channel_creds": [{"type": "insecure"}]
-		}]`, mgmtServer.Address)),
+			 "server_uri": %q,
+			 "channel_creds": [{"type": "insecure"}]
+		 }]`, mgmtServer.Address)),
```
It looks like it was indented properly earlier and this change is messing up the indenting.

purnesh42H replied:
Reverted

easwars commented on xds/internal/xdsclient/tests/cds_watchers_test.go:
```diff
@@ -67,22 +67,22 @@ func newClusterWatcher() *clusterWatcher {
 	// When used with a go-control-plane management server that continuously
 	// resends resources which are NACKed by the xDS client, using a `Replace()`
-	// here and in OnResourceDoesNotExist() simplifies tests which will have
+	// here and in ResourceError() simplifies tests which will have
```
Nit: here and in AmbientError

easwars commented on xds/internal/xdsclient/tests/eds_watchers_test.go:
```diff
@@ -76,22 +76,22 @@ func newEndpointsWatcher() *endpointsWatcher {
 	// When used with a go-control-plane management server that continuously
 	// resends resources which are NACKed by the xDS client, using a `Replace()`
-	// here and in OnResourceDoesNotExist() simplifies tests which will have
+	// here and in ResourceError() simplifies tests which will have
```
Same here.

purnesh42H replied:
Done

easwars commented on xds/internal/xdsclient/tests/lds_watchers_test.go:
```diff
@@ -71,22 +71,22 @@ func newListenerWatcher() *listenerWatcher {
 	// When used with a go-control-plane management server that continuously
 	// resends resources which are NACKed by the xDS client, using a `Replace()`
-	// here and in OnResourceDoesNotExist() simplifies tests which will have
+	// here and in ResourceError() simplifies tests which will have
```
And here.

easwars commented on xds/internal/xdsclient/tests/misc_watchers_test.go:
```diff
@@ -72,25 +72,25 @@ func newTestRouteConfigWatcher(client xdsclient.XDSClient, name1, name2 string)
 	// When used with a go-control-plane management server that continuously
 	// resends resources which are NACKed by the xDS client, using a `Replace()`
-	// here and in OnResourceDoesNotExist() simplifies tests which will have
+	// here and in ResourceError() simplifies tests which will have
```
Same here.

purnesh42H replied:
Done

easwars commented on xds/internal/xdsclient/tests/rds_watchers_test.go:
```diff
@@ -66,22 +66,22 @@ func newRouteConfigWatcher() *routeConfigWatcher {
 	// When used with a go-control-plane management server that continuously
 	// resends resources which are NACKed by the xDS client, using a `Replace()`
-	// here and in OnResourceDoesNotExist() simplifies tests which will have
+	// here and in ResourceError() simplifies tests which will have
```
And here.

purnesh42H replied:
Done

easwars commented on xds/internal/xdsclient/tests/resource_update_test.go:
```diff
@@ -161,7 +161,7 @@ func (s) TestHandleListenerResponseFromManagementServer(t *testing.T) {
 				}},
 			},
-			wantErr: "Listener not found in received response",
+			wantErr: fmt.Sprintf("xds: resource \"%v\" of type \"ListenerResource\" does not exist", resourceName1),
```
Would using `%q` instead of having to do `\"%v\"` make this easier to read?

Same with `\"ListenerResource\"`. You could do a `%q` in the format string and include "ListenerResource" as one of the arguments to `fmt.Sprintf`.

purnesh42H replied:
yeah that seem to work. Thanks.

easwars commented on xds/internal/xdsclient/xdsresource/resource_type.go:
```diff
@@ -52,33 +52,30 @@ type Producer interface {
+// received by the xDS client from the management server.
+//
+// All methods contain a done parameter which should be called when processing
+// of the update has completed.  For example, if processing a resource requires
+// watching new resources, those watches should be completed before done is
+// called, which can happen after the ResourceWatcher method has returned.
```
Could you please simplify/clarify/rephrase this sentence:

// For example, if processing a resource requires
// watching new resources, those watches should be completed before done is
// called, which can happen after the ResourceWatcher method has returned.

Specifically, what do you mean by "those watches should be completed before done is called"? I believe, just  registration of the watch, right? So, that needs to be clarified.

And I don't understand what "which can happen after the ResourceWatcher method has returned" this means.

purnesh42H replied:
I think this means that done can be called after the ResourceWatcher method that's calling the callback has returned? So, done has to be called for client to start sending new notifications.

purnesh42H replied:
I have modified the watches completion part to be more clear.

easwars commented:
Yes, I agree that we should not have checks for "resource-not-found" errors in the individual watchers. In fact, I'm thinking we might not even need those error types like `xdsresource.ErrorTypeResourceNotFound` anymore. If `ResourceError` is invoked, then the watcher must stop using the previous configuration and if `AmbientError` is invoked, it should continue using the previous configuration and the xDS client should guarantee that `AmbientError` is only invoked when a previous copy of the resource was delivered via `ResourceChanged`.

easwars commented:
> Everything else is working fine but looks like this TestServeAndCloseDoNotRace is causing race [grpc/grpc-go/actions/runs/14091957805/job/39470559995?pr=7977](https://github.com/grpc/grpc-go/actions/runs/14091957805/job/39470559995?pr=7977). Server is stopping in the middle of the test.

Sorry, I don't have the cycles to debug this failure. But if you can dig a little deeper and get more info on the sequence of events leading to the failure, and *why* exactly the flake is happening, I might be able to help better. Once we have that information, we will be able to decide if we need to rewrite the test or if there is something wrong we are doing in the xDS client. Thanks.

purnesh42H commented:
> > yeah handled it in ListenerWrapper
> 
> I haven't looked at the code, but that seems wrong to me. There should not need to be any special handling for individual resource types -- the logic for this in XdsClient should be completely resource-type-agnostic.
> 
> The only thing that should be handled in individual resource types are things that are different from other resource types, like determining what resource type to pass to the watcher.

@markdroth the case we are looking here is actually channel creation failure. In c++ and java that's never expected but in go recently a change was added to report a transient error if channel creation failed. Now, based on A88, this should be a non-ambient error only if we don't have a cached resource otherwise it should be an ambient error. Is that the correct behavior?

markdroth commented:
> @markdroth the case we are looking here is actually channel creation failure. In c++ and java that's never expected but in go recently a change was added to report a transient error if channel creation failed. Now, based on A88, this should be a non-ambient error only if we don't have a cached resource otherwise it should be an ambient error. Is that the correct behavior?

Yes.  But in practice I suspect this case will never happen when we have a cached resource, because the only cases where channel creation fails should be things like invalid target URIs, in which case the channel creation will never succeed for the life of the XdsClient, since the bootstrap config (which specifies the target URI) can't change after the XdsClient is created.  And if the channel creation never succeeds, then there's no way we can have gotten a cached resource.

The only exception I can think of to that would be fallback, if channel creation is failing for the primary server but not for the secondary server.  But in general, channel-level errors on the primary server should not be sent to the watcher in the first place if there's a secondary server to fall back to, so I think that's a non-issue.

purnesh42H commented:
@easwars the expected behavior, according to A88 is if resource is NACKed and is not cached, the watcher should drop the resource. So, in the case of xDS enabled gRPC server, if listener resource is not cached, server needs to drop the listener resource and should move to "not serving". Another thing to note is we don't suppress not_serving->not_serving  transition and invoke the mode change callback each time.

So, i have modified the server tests cases accordingly,
- TestServer_Security_NoCertificateProvidersInBootstrap_FailureTest
- Server_Security_WithValidAndInvalidSecurityConfiguration
In these, we need to drain the modeCh and errorCh channel of test mode change callback each time serving mode is switched but also make sure that it is always moving to "not serving".

For, TestResolverBadServiceUpdate, I split into 2 parts. 1) without cached 2) with cached

For, TestServeAndCloseDoNotRace, override the default ServingModeCallback because the listener resource is with invalid uri which is always expected to cause resource error leading to serving mode change. And default ServingModeCallback logs an error in that case. For this test that's irrelevant.

With this I think we have verified how the existing functionality is affected. I will also import this PR in google3 and make sure all the prod resolvers including directpath are working as expected.

easwars commented:
I don't understand what you mean by dropping the resource? Why should the listener watcher drop the resource if it sees the resource being NACKed when it was not cached previously? It should continue watching that resource in the hopes that a subsequent update for the same resource will be ACKed.

purnesh42H commented:
> I don't understand what you mean by dropping the resource? Why should the listener watcher drop the resource if it sees the resource being NACKed when it was not cached previously? It should continue watching that resource in the hopes that a subsequent update for the same resource will be ACKed.

oh by dropping the resource it doesn't mean stop watching the resource. It just means that if the consumer of watcher was using the resource, they should stop using it. So, in the case of server it leads to serving mode change to "not serving" but watcher will still have the watch on.

easwars commented:
> the expected behavior, according to A88 is if resource is NACKed and is not cached, the watcher should drop the resource

If we are talking about the case where a NACK is seen without the resource being cached, then the question of not using the resource anymore does not come into the picture, because the resource does not exist yet.

purnesh42H commented:
> > Yes, I agree that we should not have checks for "resource-not-found" errors in the individual watchers. In fact, I'm thinking we might not even need those error types like `xdsresource.ErrorTypeResourceNotFound` anymore. If `ResourceError` is invoked, then the watcher must stop using the previous configuration and if `AmbientError` is invoked, it should continue using the previous configuration and the xDS client should guarantee that `AmbientError` is only invoked when a previous copy of the resource was delivered via `ResourceChanged`.
> 
> While making another pass through this, I really think we should get rid of these error types, or at least look into getting rid of them. The reason behind my proposal is that when a watcher gets a `ResourceError`, it should simply stop using that resource if it already had it. There should be no logic of checking the error type and doing different things. Having these error types encourages that behavior.
> 
> We don't have to make that change as part of this PR, but I think you should add it to your list of things that you have that need to be done as part of the generic xDS client changes.

yes, I have added. Thanks.

purnesh42H commented:
> > the expected behavior, according to A88 is if resource is NACKed and is not cached, the watcher should drop the resource
> 
> If we are talking about the case where a NACK is seen without the resource being cached, then the question of not using the resource anymore does not come into the picture, because the resource does not exist yet.

yeah if watcher has not seen the valid resource before then its not applicable. The gRFC talks about case when watcher has seen the valid resource but it deleted in xDS client cache https://github.com/grpc/proposal/blob/master/A88-xds-data-error-handling.md#changes-to-xdsclient-watcher-apis

easwars commented on xds/internal/server/listener_wrapper.go:
```diff
@@ -441,27 +441,27 @@ func (lw *ldsWatcher) OnUpdate(update *xdsresource.ListenerResourceData, onDone
+	if lw.logger.V(2) {
+		lw.logger.Infof("LDS watch for resource %q reported ambient error: %v", lw.name, err)
+	}
+	// For errors which are anything other than "resource-not-found", we
+	// continue to use the old configuration.
```
I think we can remove this comment now since the API is clear about which error callback means what.

purnesh42H replied:
Removed.

easwars commented on xds/internal/server/listener_wrapper.go:
```diff
@@ -441,27 +441,27 @@ func (lw *ldsWatcher) OnUpdate(update *xdsresource.ListenerResourceData, onDone
-	err := xdsresource.NewErrorf(xdsresource.ErrorTypeResourceNotFound, "resource name %q of type Listener not found in received response", lw.name)
-	lw.parent.onLDSResourceDoesNotExist(err)
+	if lw.logger.V(2) {
+		lw.logger.Infof("LDS watch for resource %q reported ambient error: %v", lw.name, err)
```
Could you please check if we should include this error in the new `resolution_note` field that is being added as part of A74? If that is the case, it would be nice if you can file an issue and link that here with a TODO. Thanks.

easwars commented on xds/internal/server/listener_wrapper.go:
```diff
@@ -441,27 +441,27 @@ func (lw *ldsWatcher) OnUpdate(update *xdsresource.ListenerResourceData, onDone
 	}
-	// For errors which are anything other than "resource-not-found", we
-	// continue to use the old configuration.
+	lw.parent.onLDSResourceError(err)
```
This is the only callsite for this method. Maybe we can get rid of it?

easwars commented on xds/internal/server/listener_wrapper.go:
```diff
@@ -412,10 +412,10 @@ func (l *listenerWrapper) annotateErrorWithNodeID(err error) error {
+func (l *listenerWrapper) onLDSResourceError(err error) {
 	l.mu.Lock()
 	defer l.mu.Unlock()
-	l.switchModeLocked(connectivity.ServingModeNotServing, l.annotateErrorWithNodeID(err))
```
Also, you could get rid of the `annotateErrorWithNodeID` method since there is a single caller for it now. So, the annotation can be inlined at the callsite.

purnesh42H replied:
Done

easwars commented on xds/internal/server/rds_handler.go:
```diff
@@ -165,30 +164,28 @@ func (rw *rdsWatcher) OnUpdate(update *xdsresource.RouteConfigResourceData, onDo
-	if rw.logger.V(2) {
-		rw.logger.Infof("RDS watch for resource %q reported error: %v", rw.routeName, err)
-	}
+	rw.logger.Warningf("RDS watch for resource %q reported resource error", rw.routeName)
```
Same here too. Info and v2.

easwars commented on xds/internal/server/rds_handler.go:
```diff
@@ -165,30 +164,28 @@ func (rw *rdsWatcher) OnUpdate(update *xdsresource.RouteConfigResourceData, onDo
-		rw.logger.Infof("RDS watch for resource %q reported error: %v", rw.routeName, err)
-	}
+	rw.logger.Warningf("RDS watch for resource %q reported resource error", rw.routeName)
 	rw.parent.handleRouteUpdate(rw.routeName, rdsWatcherUpdate{err: err})
```
All callbacks are now calling the `handleRouteUpdate` method on the parent, which is looking at the error type and deciding how to handle it. That is not what we want, I believe. We need to handle resource errors and ambient errors differently. So, maybe we need different callbacks on the parent, or there is a better way to do it?

purnesh42H replied:
Removed handleRouteUpdate. Updated the 3 watcher methods to do the expected things.

easwars commented on xds/server_test.go:
```diff
@@ -655,15 +655,13 @@ func (s) TestHandleListenerUpdate_ErrorUpdate(t *testing.T) {
-	// requested resource is not present in the management server. This happens
-	// when the watch timer expires or when the resource is explicitly deleted
-	// by the management server.
+	// Also make sure that serving mode updates are received. The serving
+	// mode changes to NOT_SERVING. This happens because watcher received an
+	// invalid resource from the server which is not present in cache.
```
The cache is present in the xDS client. So, I would recommend updating this comment to just say that "This happens because the watcher received a resource error from the xDS client".

purnesh42H replied:
Done

easwars commented on xds/server_test.go:
```diff
@@ -697,10 +695,23 @@ func (s) TestServeAndCloseDoNotRace(t *testing.T) {
 	// Generate bootstrap contents up front for all servers.
 	bootstrapContents := generateBootstrapContents(t, uuid.NewString(), nonExistentManagementServer)
 
+	// Override the default ServingModeCallback with a noop function because the
+	// invalid listener resource will be immediately NACKed by the xDS client
+	// and since the listener resource is not cached, it will trigger multiple
+	// resource error notifications for the same listener resource in quick
+	// successions, leading to service mode change to "not serving" each time.
```
I don't quite understand this. We don't even have a management server in this test. So, where does the question of an invalid listener resource arise?

Is the xdsChannel creation to this management server failing? If so, why?

purnesh42H replied:
yes, xdsChannel creation failure. I have updated the comment.

easwars commented on xds/server_security_ext_test.go:
```diff
@@ -128,7 +128,8 @@ func (s) TestServer_Security_NoCertProvidersInBootstrap_Success(t *testing.T) {
 // bootstrap file. The test verifies that server creation does not fail and that
-// the xDS-enabled gRPC server does not enter "serving" mode.
+// the xDS-enabled gRPC server does not enter "serving" mode but serving mode
+// changes to "not serving" because the listener resource is not cached.
```
The serving mode changing to "not serving" is not technically true. The server does not start in "serving mode". When the server is created and it receives its configuration, it enters either "serving" or "not serving" mode (and can then switch modes during its lifetime based on other configuration updates).

purnesh42H replied:
Updated

easwars commented on xds/server_security_ext_test.go:
```diff
@@ -214,25 +215,38 @@ func (s) TestServer_Security_NoCertificateProvidersInBootstrap_Failure(t *testin
+	exit := false
+	for {
+		select {
+		case <-ctx.Done():
+			t.Fatalf("Timed out waiting for server to go NOT_SERVING")
+		case gotMode := <-modeChangeHandler.modeCh:
+			if gotMode != connectivity.ServingModeNotServing {
+				t.Fatalf("Mode changed to %v, want %v", gotMode, connectivity.ServingModeNotServing)
+			}
+			gotErr := <-modeChangeHandler.errCh
+			if gotErr == nil || !strings.Contains(gotErr.Error(), nodeID) {
+				t.Fatalf("Unexpected error: %v, want xDS Node id: %s", gotErr, nodeID)
+			}
+		default:
+			exit = true
```
This is racy. You cout hit the "default" case for the `select` the first time around when your server has still not switched the mode to "not serving" and the test will pass incorrectly.

purnesh42H replied:
Update the test handler to suppress error updates for no_serving -> no_serving cases.

easwars commented on xds/server_security_ext_test.go:
```diff
@@ -214,25 +215,38 @@ func (s) TestServer_Security_NoCertificateProvidersInBootstrap_Failure(t *testin
+	// at the same time verify that the server is NOT_SERVING and the error
+	// message contains the xDS node ID.
+	exit := false
+	for {
```
Can we suppress the mode change callback if the mode is staying in NOT_SERVING and the error is the same? This shouldn't happen in practice a lot. But we never know what kind of management servers people run in the wild. If we do suppress these duplicates, then we wont need this `for` loop here.

purnesh42H replied:
Update the test handler to suppress error updates for no_serving -> no_serving cases.

easwars replied:
Actually my comment was to suppress calling the mode change callback from the listener wrapper in the server, and not suppressing the event once the callback has been invoked. Could you please add a TODO here or in your other tracker to do that.

easwars replied:
Yes, we can check what the other languages do and discuss what would be the best way forward.

easwars commented on xds/server_security_ext_test.go:
```diff
@@ -214,25 +215,38 @@ func (s) TestServer_Security_NoCertificateProvidersInBootstrap_Failure(t *testin
 	}
-	defer cc.Close()
-
-	waitForFailedRPCWithStatus(ctx, t, cc, codes.Unavailable, "", "")
```
We should still be checking for RPC failures with the correct status and xDS node ID in the error string.

purnesh42H replied:
yes, added back.

easwars commented on xds/server_security_ext_test.go:
```diff
@@ -245,7 +259,8 @@ func (s) TestServer_Security_NoCertificateProvidersInBootstrap_Failure(t *testin
 //
 // The test verifies that an RPC to the first listener succeeds, while the
-// second listener never moves to "serving" mode and RPCs to it fail.
+// second listener never moves to "serving" mode and RPCs but serving mode
```
Same comment applies to this test as the previous one.

purnesh42H replied:
Done

purnesh42H commented on xds/internal/xdsclient/clientimpl_watchers.go:
```diff
@@ -61,15 +65,15 @@ func (c *clientImpl) WatchResource(rType xdsresource.Type, resourceName string,
 
 	if err := c.resourceTypes.maybeRegister(rType); err != nil {
 		logger.Warningf("Watch registered for name %q of type %q which is already registered", rType.TypeName(), resourceName)
```
Done

purnesh42H commented on xds/internal/xdsclient/tests/authority_test.go:
```diff
@@ -365,11 +365,11 @@ func (cw *clusterWatcherV2) OnError(err error, onDone xdsresource.OnDoneFunc) {
 	// here simplifies tests that want access to the most recently received
 	// error.
-	cw.resourceNotFoundCh.Replace(xdsresource.NewError(xdsresource.ErrorTypeResourceNotFound, "Cluster not found in received response"))
+	cw.resourceNotFoundCh.Replace(err)
```
yes, updated.

purnesh42H commented on xds/internal/testutils/resource_watcher.go:
```diff
@@ -31,40 +31,43 @@ type TestResourceWatcher struct {
 	UpdateCh chan *xdsresource.ResourceData
 	// ErrorCh is the channel on which errors from the xDS client are delivered.
 	ErrorCh chan error
-	// ResourceDoesNotExistCh is the channel used to indicate calls to OnResourceDoesNotExist
```
Updated

purnesh42H replied:
Done

purnesh42H commented on xds/internal/balancer/cdsbalancer/cdsbalancer.go:
```diff
@@ -357,7 +357,7 @@ func (b *cdsBalancer) ResolverError(err error) {
 			root = b.lbCfg.ClusterName
 		}
-		b.onClusterError(root, err)
+		b.onClusterAmbientError(root, err)
```
updated as discussed. Ambient error will only pass the error to child policy. Resource error will always close child policy and report TF. In case of updates from resolver, we decide to call ambient or resource error based on presence of child policy.

purnesh42H commented on xds/internal/balancer/cdsbalancer/cdsbalancer.go:
```diff
@@ -566,14 +566,13 @@ func (b *cdsBalancer) onClusterError(name string, err error) {
-// TRANSIENT_FAILURE.
+// Handles an error Cluster update from the xDS client to stop using the
+// previously seen resource. Propagates the error down to the child policy
+// if one exists, or puts the channel in TRANSIENT_FAILURE.
```
Updated

purnesh42H commented on xds/internal/balancer/clusterresolver/resource_resolver_eds.go:
```diff
@@ -89,52 +89,52 @@ func (er *edsDiscoveryMechanism) OnUpdate(update *xdsresource.EndpointsResourceD
 
-	// Report an empty update that would result in no priority child being
+	er.mu.Lock()
+	if er.update != nil {
```
yes, no need to do anything else here.

purnesh42H commented on xds/internal/balancer/clusterresolver/resource_resolver_eds.go:
```diff
@@ -89,52 +89,52 @@ func (er *edsDiscoveryMechanism) OnUpdate(update *xdsresource.EndpointsResourceD
-		onDone()
-		return
-	}
+	er.logger.Warningf("EDS discovery mechanism for resource %q reported resource error: %v", er.nameToWatch, err)
```
Done

purnesh42H commented on xds/internal/balancer/clusterresolver/resource_resolver_eds.go:
```diff
@@ -89,52 +89,52 @@ func (er *edsDiscoveryMechanism) OnUpdate(update *xdsresource.EndpointsResourceD
 }
 
-func (er *edsDiscoveryMechanism) OnError(err error, onDone xdsresource.OnDoneFunc) {
+func (er *edsDiscoveryMechanism) ResourceError(err error, onDone func()) {
```
ResourceError actually sends an empty update but AmbientError only logs. Post 74 this resolver will be removed https://github.com/grpc/proposal/blob/master/A74-xds-config-tears.md#removing-the-xds_cluster_resolver-lb-policy so may be we just keep it as no-op for now?

purnesh42H commented on xds/internal/balancer/clusterresolver/resource_resolver_eds.go:
```diff
@@ -89,52 +89,52 @@ func (er *edsDiscoveryMechanism) OnUpdate(update *xdsresource.EndpointsResourceD
+		return
+	}
+
+	// Else report an empty update that would result in no priority child being
```
Actually, we don't have to update in case of ambient error because we are sure that previous update exist. So, i removed it.

purnesh42H commented on xds/internal/resolver/xds_resolver.go:
```diff
@@ -546,16 +546,16 @@ func (r *xdsResolver) onListenerResourceUpdate(update xdsresource.ListenerUpdate
-func (r *xdsResolver) onListenerResourceNotFound() {
-	r.logger.Warningf("Received resource-not-found-error for Listener resource %q", r.ldsResourceName)
+func (r *xdsResolver) onListenerResourceError(err error) {
+	r.logger.Warningf("Received resource error for Listener resource %q: %v", r.ldsResourceName, err)
```
Done

purnesh42H commented on xds/internal/resolver/xds_resolver.go:
```diff
@@ -584,16 +584,16 @@ func (r *xdsResolver) onRouteConfigResourceUpdate(name string, update xdsresourc
-func (r *xdsResolver) onRouteConfigResourceNotFound(name string) {
-	r.logger.Warningf("Received resource-not-found-error for RouteConfiguration resource %q", name)
+func (r *xdsResolver) onRouteConfigResourceError(name string, err error) {
+	r.logger.Warningf("Received resource error for RouteConfiguration resource %q: %v", name, err)
```
Thanks. Done.

purnesh42H commented on xds/internal/resolver/xds_resolver_test.go:
```diff
@@ -288,13 +288,10 @@ func (s) TestResolverCloseClosesXDSClient(t *testing.T) {
-func (s) TestResolverBadServiceUpdate(t *testing.T) {
+// Tests the case where a resource, not present in cache, returned by the
+// management server is NACKed by the xDS client, which then returns an update
+// containing a resource error to the resolver.
```
Added

purnesh42H commented on xds/internal/resolver/xds_resolver_test.go:
```diff
@@ -319,12 +316,29 @@ func (s) TestResolverBadServiceUpdate(t *testing.T) {
 
-	// Build the resolver and expect an error update from it.
-	stateCh, errCh, _ := buildResolverForTarget(t, resolver.Target{URL: *testutils.MustParseURL("xds:///" + defaultTestServiceName)}, bc)
-	wantErr := "no RouteSpecifier"
```
Made the change passthrough the resource error all the way. Before it was throwing static error with "no valid clusters"

purnesh42H commented on xds/internal/resolver/xds_resolver.go:
```diff
@@ -584,16 +584,16 @@ func (r *xdsResolver) onRouteConfigResourceUpdate(name string, update xdsresourc
+	r.logger.Warningf("Received resource error for RouteConfiguration resource %q: %v", name, err)
 
 	if r.rdsResourceName != name {
 		return
```
Renamed and made change to passthrough error all the way.

purnesh42H commented on xds/internal/resolver/xds_resolver_test.go:
```diff
@@ -335,10 +349,27 @@ func (s) TestResolverBadServiceUpdate(t *testing.T) {
 
-	// Configure another bad resource on the management server and expect an
-	// error update from the resolver.
+	// Configure a listener resource that is expected to be NACKed because it
```
Done

purnesh42H commented on xds/internal/resolver/xds_resolver_test.go:
```diff
@@ -335,10 +349,27 @@ func (s) TestResolverBadServiceUpdate(t *testing.T) {
-	if err := waitForErrorFromResolver(ctx, errCh, wantErr, nodeID); err != nil {
+	if err := waitForErrorFromResolver(ctx, errCh, "no RouteSpecifier", nodeID); err != nil {
 		t.Fatal(err)
 	}
```
Done

purnesh42H commented on xds/server_ext_test.go:
```diff
@@ -169,7 +169,8 @@ func waitForFailedRPCWithStatus(ctx context.Context, t *testing.T, cc *grpc.Clie
 			_, err = client.EmptyCall(ctx, &testpb.Empty{})
-			if gotCode := status.Code(err); gotCode != wantCode {
+			gotCode := status.Code(err)
+			if gotCode != wantCode {
```
Done

easwars commented on xds/internal/resolver/xds_resolver.go:
```diff
@@ -482,7 +482,7 @@ func (r *xdsResolver) onError(err error) {
 //
 // Only executed in the context of a serializer callback.
-func (r *xdsResolver) onResourceNotFound() {
+func (r *xdsResolver) onResourceError(err error) {
```
Github doesn't allow me to comment on unchanged lines. But in method `applyRouteConfigUpdate`, there is a call to `onError`. I feel that this is also a data error and also belongs to the class of error I was talking about in the comment in `xds/internal/balancer/cdsbalancer/cdsbalancer.go`.

purnesh42H replied:
yeah that's kind of weird. I have added the TODO. Its again being called from resource update method which has finished applying the new update. For now, I have kept it as ambient error to not break existing behavior.

easwars replied:
Please link a github issue with full details in any TODO that is being added, unless you plan to act on the TODO right away in which case you can add your github handle to the todo. Thanks.

purnesh42H replied:
Added alias. It will be fast follow from okr tracker

easwars commented on xds/server_ext_test.go:
```diff
@@ -97,10 +100,22 @@ func newServingModeChangeHandler(t *testing.T) *servingModeChangeHandler {
+	m.currentMode = args.Mode
 	if args.Err != nil {
 		m.errCh <- args.Err
+		m.currentErr = args.Err
```
This field should be updated unconditionally, shouldn't it?

purnesh42H replied:
yeah I was updating error to nil if mode was changing to "serving" and updating error only if it was nil. But unconditional update is simple as in case of mode change to "serving", err will be `nil`

easwars commented on xds/internal/server/rds_handler.go:
```diff
@@ -174,21 +162,28 @@ func (rw *rdsWatcher) OnError(err error, onDone xdsresource.OnDoneFunc) {
+	}
+	routeName := rw.routeName
+	rwu := rw.parent.updates[routeName]
+	rw.parent.updates[routeName] = rwu
```
This line is a no-op. Can we delete it?

purnesh42H replied:
Done

easwars commented on xds/internal/server/listener_wrapper.go:
```diff
@@ -438,30 +395,62 @@ func (lw *ldsWatcher) OnUpdate(update *xdsresource.ListenerResourceData, onDone
+	// What this means is that the XDSClient has ACKed a resource which can push
+	// the server into a "not serving" mode. This is not ideal, but this is
+	// what we have decided to do.
+	if ilc.Address != l.addr || ilc.Port != l.port {
```
This also belongs to the same class of errors I mentioned in the cds LB policy and the xDS resolver.

purnesh42H replied:
Added a TODO. Again this is withing ResourceChanged so management server wants us to drop the resource?

easwars reviewed (commented):
Mostly LGTM. Very minor comments this time around. But some action items are there. So, please let's make sure we have them tracked. Thanks.

easwars commented on xds/internal/balancer/cdsbalancer/cdsbalancer.go:
```diff
@@ -474,20 +478,28 @@ func (b *cdsBalancer) onClusterUpdate(name string, update xdsresource.ClusterUpd
 			// instance is not found in the bootstrap config, we need to put the
 			// channel in transient failure.
-			b.onClusterError(name, b.annotateErrorWithNodeID(fmt.Errorf("received Cluster resource contains invalid security config: %v", err)))
+			b.onClusterResourceError(name, b.annotateErrorWithNodeID(fmt.Errorf("received Cluster resource contains invalid security config: %v", err)))
```
Also, the same code of checking if the child policy exists or not and calling one of the onClusterXxxError methods is repeated three times. We could pull it into a helper method.

purnesh42H replied:
Done
